### DISCUSSION TITLE
MemTest86 UEFI unable to submit reports or status using TFTP when chain loaded from iPXE

### DISCUSSION THREAD

**Discussion by rushyn** [ipxe/ipxe#868] [General]
MemTest86 Pro Site Edition is PXE bootable and is able to push test status and final report to a TFTP server.  This works then MemTest is started directly my PXE. But then chain loaded from iPXE its not able to push reports to TFTP server. Its still able to pull the config file from TFTP but its not able to put files on the TFTP server.

Is there a workaround for this?

**Comment by mcb30** [comment]
Could you try building with `DEBUG=pxe_udp,pxe_tftp` to see what errors occur when you attempt to upload a file to the TFTP server?

**Reply by mcb30** [reply]
Sorry, just noticed you said UEFI. Ignore the previous message!

**Comment by mcb30** [comment]
iPXE doesn't have support for TFTP write in UEFI: https://github.com/ipxe/ipxe/blob/2d180ce233f6eae8d0df279e5cbf134f547f604c/src/interface/efi/efi_pxe.c#L950-L957

This is largely because I've never previously come across anything that uses UEFI TFTP write, and so there was nothing to test against.

It would be a moderate chunk of work to add, so is unlikely to happen unless a commercial organisation decides to fund it.

**Reply by mcb30** [reply]
As a workaround: if this version of memtest is able to exit with an exit status that reflects success/failure then you can catch that in the calling iPXE script and use it to report the status another way, e.g. via an HTTP POST request

**Reply by NiKiZe** [reply]
Any chance to build iPXE without iPXEs TFTP implementation and only use firmware?
Or maybe passthru write to firmware?

**Reply by mcb30** [reply]
> Any chance to build iPXE without iPXEs TFTP implementation and only use firmware? Or maybe passthru write to firmware?

It could be done but it would be quite messy, and leave us with something that's hard to debug if it goes wrong. It's almost certainly less work overall to implement TFTP write support in iPXE.

**Comment by rushyn** [comment]
@mcb30 Thank you for looking in to this, its the answerer I expected. I do not think I can get get the company I work for to spend any money on this as this is not something we need. Feel free to let me know how much you think it will cost to add TFTP put functionality but as I said, its unlikely I will be able to get it justified and approved.

I can see what you are saying about capturing the exit code, I don't know if MemTest86 UEFI has an exit code that indicates pass or fail, but its something I can find out. However I am no so match after the exit code as to record pass fail as the report file it self. I have attached the report files so that you can see want I mean.

This was originally a HTML file uploaded by MemTest86 to TFTP.

![BSN12345678901234567-MemTest86-Report-20230127-180346](https://user-images.githubusercontent.com/65286397/215039785-b9a25f11-d35f-4ebc-9632-016ae58a9d2d.jpg)

And this was originally a bit map file.
![DIMMResults-20230127-180400](https://user-images.githubusercontent.com/65286397/215039929-a091c03a-2af7-46c7-97bc-62b06fe92796.jpg)

The test status is updated by MemTest86 by uploading an XML file to TFTP every minute. 

![image](https://user-images.githubusercontent.com/65286397/215040625-f29ba3db-8b3f-4909-afbb-c188c7d77824.png)

**Reply by mcb30** [reply]
It does look like a neat feature to have in iPXE! You're very welcome to try to persuade Passmark to provide a free Site Edition licence for the iPXE project to use - that should be enough to get me to implement it in my own time, since it's an interesting thing to add.

It would have to be the Site Edition licence, I think - it sounds as though the Pro Edition is tied to a specific client machine, which would interfere with my usual process of testing across multiple machines and disposable VMs.

**Reply by rushyn** [reply]
@mcb30 Pro Edition only works of USB drive, Then you get Site Edtion you get access to both version the PXE version and the USB veriosn. You can't PXE boot PRO version I think there is an actual block that stops you, I will check this on monday. 

On monday I will email passmark and expaline what you are want to do. There support has been good then we had issues. But I do not know what they are going to say regarding a free license.  They are aware of the issue with iPXE as I had try to see if this could be resolved from there end two years ago.

memtest86.com/tech_pxe-client-config.html
![image](https://user-images.githubusercontent.com/65286397/215062584-8b18d540-aa89-400e-8adc-6c1a6ca6d65f.png)

**Reply by mcb30** [reply]
Thank you. Please feel free to CC me in on the email. I suspect they'll be more responsive to a request coming from you (a customer) than if I were to email them myself.

**Reply by rushyn** [reply]
@mcb30 I should use the email in your github profile then I cc you?

**Reply by mcb30** [reply]
> @mcb30 I should use the email in your github profile then I cc you?

Yes, that's fine, thanks

**Reply by NX-06** [reply]
Hello,
@mcb30 has there been any progress on this feature?

**Reply by mcb30** [reply]
@NX-06 No, I've been pretty short on free time for the last year, sorry

**Reply by NX-06** [reply]
No worries, thanks for letting me know.

**Comment by rushyn** [comment]
@mcb30 I received an email from Passmark, It said that they have sent software for testing, can you confirm that you have received it?

**Reply by mcb30** [reply]
Yes, I have received it this morning.  Thank you for your help in arranging this!
